Fix #7002: Always invalidate virtual floor regardless of selection.

ZehMatt opened the pull request:
Currently the virtual floor would not invalidate if placement would be impossible, this should correct the issue.
![2018-03-03_12-36-44](https://user-images.githubusercontent.com/5415177/36933821-ab95fcac-1edf-11e8-89a0-034b4376936e.gif)

AaronVanGeffen commented:
Thanks, this appears to fix the horizontal invalidation problems nicely. It does not fix the redrawing lag, but that's a different issue, of course.

ZehMatt commented:
What would be the best way to reproduce the redraw lag?

AaronVanGeffen commented:
Move the selection up and down quickly. You'll notice part of the floor is lagging behind. For me, it's usually the centre not being redrawn, while the outer parts are. It's only there for a fraction of a second, but noticeable nonetheless.

ZehMatt commented:
I can only get it slightly to lag with software rendering not with opengl, could you please test that on your end?

AaronVanGeffen commented:
Yeah, it only happens with the software renderer. OpenGL does not have the problem, as it doesn't use invalidation.

ZehMatt commented:
Makes sense, at least I know where to look now, thank you.

ZehMatt commented:
This should do. Would be nice if someone can confirm it.

AaronVanGeffen commented on src/openrct2/world/Map.cpp:
```diff
@@ -4819,11 +4819,6 @@ void map_invalidate_virtual_floor_tiles()
 
-    // Do not invalidate if we're between ticks.
-    if (min_position.x == std::numeric_limits<sint16>::max()    || min_position.y == std::numeric_limits<sint16>::max() ||
-        max_position.x == std::numeric_limits<sint16>::lowest() || max_position.y == std::numeric_limits<sint16>::lowest())
```
This was some very deliberate coding — removing it has a very real performance impact. Without it, FPS drops dramatically with the virtual floor active: down to 10fps (active) from 70fps (inactive) here.

ZehMatt replied:
But does it actually solve the issue? I'll have a look at the performance issue, I haven't seen it on my end.

AaronVanGeffen replied:
Yes, removing it does solve the issue with rendering… If you could achieve that without the dip in FPS, I'll happily merge.

ZehMatt commented:
I've moved most of the virtual floor code to VirtualFloor.cpp and changed the way it invalidates the tiles. I will clean up the code once I know all of it is working properly now.

AaronVanGeffen commented:
You've hit the nail on the head this time — it's faster than it's ever been, for me, too. Cheers. If you could clean up the transformation functions, I'll be happy to merge this. (You can make the originals non-static and export them through Map.h, if you like.)

AaronVanGeffen commented on src/openrct2/paint/VirtualFloor.cpp:
```diff
@@ -30,6 +45,181 @@ static constexpr const CoordsXY offsets[4] =
+
+void virtual_floor_enable()
+{
+    if (virtual_floor_is_enabled() == true)
```
Now returns a bool, so you can leave out the ` == true`.

AaronVanGeffen commented on src/openrct2/paint/VirtualFloor.cpp:
```diff
@@ -30,6 +45,181 @@ static constexpr const CoordsXY offsets[4] =
+
+void virtual_floor_disable()
+{
+    if (virtual_floor_is_enabled() == false)
```
Now returns a bool, so you can leave out the ` == false`.
